README.md
# 魔筷大前端 vscode 插件:mk-vscode-tool



vscode 插件目前支持的功能是后台项目开发，常用代码片段的支持,后期会添加新的功能。

## 用法

直接键入前缀 mk-xxx,箭头切换就可以预览代码


## 目前支持的 snippets 汇总

这个表格是我们开发后台使用频率比较高的代码



|序号| 前缀                  |                        代码                         |
|:----:| :--------------------- | :-------------------------------------------------|
|1.| `mk-formTable`    |                    `formTable 组件代码 `             |
|2.| `mk-dialog`             |               `弹框组件代码片段 `               |
|3.| `mk-dialogForm  `       |               `elementUI 表单弹框组件 `               |
|4.| `mk-get`          |                    `get 请求代码 `                         |
|5.| `mk-post `              |                    `post 请求代码 `                   |
|6.| `mk-push`               |                    `push路由跳转`                   |
|7.| `mk-message`            |                  `elmentUI 消息提示`                  |
|8.| `mk-message-error`      |                `elmentUI 错误消息提示`                |
|9.| `mk-message-warning`    |                `elmentUI 警告消息提示 `               |
|10.| `mk-message-success`    |                `elmentUI 成功消息提示`                |
|11.| `mk-messageBox-alert`   |              `elementUI alert 消息弹框`               |
|12.| `mk-messageBox-confirm` |               `elementUI 提示消息弹框`                |
|13.| `mk-searchFromTable` |               `searchFromTable组件(已经停止维护)`                |

---



 ElementUI 部分
 
|序号 |前缀            |                  Element Tag |
|:---|:--------------|:------------------------------|
|1. | `row` | `<el-row>` |
|2. | `col` | `<el-col>` |
|3. | `elb` | `<el-button>` |

#### Form Part

|序号    |        前缀   |                Element Tag|
|:-----:|:--------------|:-------------------------------|
|1.  | `radio` | `<el-radio>` |
|2.  | `radiogroup` | `<el-radio-group>` |
|3.  | `checkbox` | `<el-checkbox>` |
|4.  | `checkboxgroup` | `<el-checkbox-group>` |
|5.  | `input` | `<el-input>` |
|6.  | `textarea` | `<el-input type="textarea">` |
|7.  | `input-number` | `------` |
|8.  | `select` | `<el-select>` |
|9.  | `option` | `<el-option>` |
|10. | `cascader` | `<el-cascader>` |
|11. | `swtich` | `<el-swtich>` |
|12. | `slider` | `-----------` |
|13. | `timepicker` | `<el-time-picker>` |
|14. | `timeselect` | `-----` |
|15. | `datepicker` | `<el-date-picker>` |
|16. | `datetimepicker` | `<el-date-picker type="datetime">` |
|17. | `upload` | `<el-upload>` |
|18. | `rate` | `<el-rate>` |
|19. | `colorpicker` | `<el-color-picker>` |
|20. | `transfer` | `<el-transfer>` |
|21. | `form` | `<el-form>` |
|22. | `formitem` | `<el-form-item>` |

#### Data Part

|序号|  前缀| Element Tag|
|:------:|:--------------|:--------|
|1. | `table` | `<el-table>` |
|2. | `tablecolumn` | `<el-table-column>` |
|3. | `tag` | `----` |
|4. | `progress` | `----` |
|5. | `progress` | `----` |
|6. | `tree` | `----` |
|7. | `pagination` | `<el-pagination>` |
|8. | `badge` | `<el-badge>` |

#### Navigation Part
|序号|  前缀 | Element Tag|
|:------:|:--------------|:--------|
|1. | `menu` | `<el-menu>` |
|2. | `submenu` | `<el-submenu>` |
|3. | `menuitem` | `<el-menu-item>` |
|4. | `tabs` | `<el-tabs>` |
|5. | `tabpane` | `<el-tab-pane>` |
|6. | `breadcrumb` | `<el-breadcrumb>` |
|7. | `breadcrumbitem` | `<el-breadcrumb-item>` |
|8. | `dropdown` | `<el-dropdown>` |
|9. | `dropdownitem` | `<el-dropdown-item>` |
|10. | `steps` | `<el-steps>` |
|11. | `step` | `<el-step>` |


#### Others Part
|序号|  前缀 | Element Tag|
|:------:|:--------------|:--------|
|1. | `dialog` | `<el-dialog>` |
|2. | `tooltip` | `<el-tooltip>` |
|3. | `popover` | `<el-popover>` |
|4. | `card` | `<el-card>` |
|5. | `carousel` | `<el-carousel>` |
|6. | `carouselitem` | `<el-carousel-item>` |
|7. | `collapse` | `<el-collapse>` |
|8. | `collapseitem` | `<el-collapse-item>` |










## Snippets

Support snippets list:

* `mk-get`

  ```
    let data = {};
    lib.api.get({
        api: '',
        data: data,
        success:(res) => {
        },
        error: (err) => {
        },
        complete: () => {
        }
    })
  ```

* `mk-post`

  ```
    let data = {};
    lib.api.post({
        api: '',
        data: data,
        success:(res) => {
        },
        error: (err) => {
        },
        complete: () => {
        }
    })
  ```

* `mk-alert`

  ```
    this.$alert('这是一段内容', '标题名称', {
    confirmButtonText: '确定',
    callback: action => {
        this.$message({
            type: 'info',
            message: `action: ${ action }`
            });
        }
    });
  ```

* `mk-confirm`

  ```
    this.$confirm('此操作将永久删除该文件, 是否继续?', '提示', {
        confirmButtonText: '确定',
        cancelButtonText: '取消',
        type: 'warning'
        }).then(() => {
            this.$message({
                type: 'success',
                message: '删除成功!'
            });
        }).catch(() => {
        this.$message({
            type: 'info',
            message: '已取消删除'
        }); 
    });
  ```

* `mk-push`

    ```
     this.$router.push({ path: `` });
    ```
* `mk-message`

    ```
    this.$message('这是一条消息提示');
    ```
* `mk-message-error`

    ```
     this.$message.error('这是一条错误消息提示');
    ```
* `mk-message-success`

    ```
    this.$message.success('这是一条成功消息提示');
    ```
* `mk-message-warning`

    ```
     this.$message.warning('这是一条警告消息提示');
    ```
* `msg`

  ```
  this.$message({
    message: '',
    type: ''
  })
  ```

* `alert`

  ```
  this.$alert('', '', {
    confirmButtonText: '',
    callback: () => {}
  });
  ```

* `confirm`

  ```
  this.$confirm('', '', {
    confirmButtonText: '',
    cancelButtonText: '',
    type: ''
  }).then(() => {})
    .catch(() => {});
  ```

* `prompt`

  ```
  this.$prompt('', '', {
    confirmButtonText: '',
    cancelButtonText: '',
    inputPattern: //,
    inputErrorMessage: ''
  }).then(({ value }) => {})
    .catch(() => {});
  ```

* `msgb`

  ```
  this.$msgbox({
    title: '',
    message: '',
    showCancelButton: '',
    confirmButtonText: '',
    cancelButtonText: '',
    beforeClose: (action, instance, done) => {}
  }).then(action => {});
  ```

* `notify`

  ```
  this.$notify({
    title: '',
    message: ''
  });
  ```
* `dialog`

  ```
  <el-dialog
    title="提示"
    :visible.sync="dialogVisible"
    width="30%"
    :before-close="handleClose">
    <span>这是一段信息</span>
    <span slot="footer" class="dialog-footer">
      <el-button @click="dialogVisible = false">取 消</el-button>
      <el-button type="primary" @click="dialogVisible = false">确 定</el-button>
    </span>
  </el-dialog>
  ```

## 写在最后

一起来顺畅的写代码!目前插件还没有很完善,有任何问题都可以来和我交流一下,期待新的想法,繁杂基础的工作都可以交给工具来处理,我们只负责核心逻辑的构建!

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "mk-vscode-tool" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/webview.js
const vscode = require('vscode');
const fs = require('fs');
const path = require('path');
const util = require('./util');

/**
 * 从某个HTML文件读取能被Webview加载的HTML内容
 * @param {*} context 上下文
 * @param {*} templatePath 相对于插件根目录的html文件相对路径
 */
function getWebViewContent(context, templatePath) {
    const resourcePath = util.getExtensionFileAbsolutePath(context, templatePath);
    const dirPath = path.dirname(resourcePath);
    let html = fs.readFileSync(resourcePath, 'utf-8');
    // vscode不支持直接加载本地资源，需要替换成其专有路径格式，这里只是简单的将样式和JS的路径替换
    html = html.replace(/(<link.+?href="|<script.+?src="|<img.+?src=")(.+?)"/g, (m, $1, $2) => {
        return $1 + vscode.Uri.file(path.resolve(dirPath, $2)).with({ scheme: 'vscode-resource' }).toString() + '"';
    });
    return html;
}

/**
 * 执行回调函数
 * @param {*} panel 
 * @param {*} message 
 * @param {*} resp 
 */
function invokeCallback(panel, message, resp) {
    console.log('回调消息：', resp);
    // 错误码在400-600之间的，默认弹出错误提示
    if (typeof resp == 'object' && resp.code && resp.code >= 400 && resp.code < 600) {
        util.showError(resp.message || '发生未知错误！');
    }
    panel.webview.postMessage({ cmd: 'vscodeCallback', cbid: message.cbid, data: resp });
}

/**
 * 存放所有消息回调函数，根据 message.cmd 来决定调用哪个方法
 */
const messageHandler = {
    // 弹出提示
    alert(global, message) {
        util.showInfo(message.info);
    },
    // 显示错误提示
    error(global, message) {
        util.showError(message.info);
    },
    // 获取工程名
    getProjectName(global, message) {
        invokeCallback(global.panel, message, util.getProjectName(global.projectPath));
    },
    openFileInFinder(global, message) {
        util.openFileInFinder(`${global.projectPath}/${message.path}`);
        // 这里的回调其实是假的，并没有真正判断是否成功
        invokeCallback(global.panel, message, { code: 0, text: '成功' });
    },
    openFileInVscode(global, message) {
        util.openFileInVscode(`${global.projectPath}/${message.path}`, message.text);
        invokeCallback(global.panel, message, { code: 0, text: '成功' });
    },
    openUrlInBrowser(global, message) {
        util.openUrlInBrowser(message.url);
        invokeCallback(global.panel, message, { code: 0, text: '成功' });
    }
};

module.exports = function (context) {

    // 注册命令，可以给命令配置快捷键或者右键菜单
    // 回调函数参数uri：当通过资源管理器右键执行命令时会自动把所选资源URI带过来，当通过编辑器中菜单执行命令时，会将当前打开的文档URI传过来
    context.subscriptions.push(vscode.commands.registerCommand('extension.demo.openWebview', function (uri) {
        // 工程目录一定要提前获取，因为创建了webview之后activeTextEditor会不准确
        const projectPath = util.getProjectPath(uri);
        if (!projectPath) return;
        const panel = vscode.window.createWebviewPanel(
            'testWebview', // viewType
            "WebView演示", // 视图标题
            vscode.ViewColumn.One, // 显示在编辑器的哪个部位
            {
                enableScripts: true, // 启用JS，默认禁用
                retainContextWhenHidden: true, // webview被隐藏时保持状态，避免被重置
            }
        );
        let global = { projectPath, panel };
        panel.webview.html = getWebViewContent(context, 'src/view/test-webview.html');
        panel.webview.onDidReceiveMessage(message => {
            if (messageHandler[message.cmd]) {
                messageHandler[message.cmd](global, message);
            } else {
                util.showError(`未找到名为 ${message.cmd} 回调方法!`);
            }
        }, undefined, context.subscriptions);
    }));
};


src/welcome.js
const vscode = require('vscode');
const util = require('./util');
const fs = require('fs');
const path = require('path');
console.log('如果设置里面开启了欢迎页显示，启动欢迎页1');
/**
 * 从某个HTML文件读取能被Webview加载的HTML内容
 * @param {*} context 上下文
 * @param {*} templatePath 相对于插件根目录的html文件相对路径
 */
function getWebViewContent(context, templatePath) {
    const resourcePath = util.getExtensionFileAbsolutePath(context, templatePath);
    const dirPath = path.dirname(resourcePath);
    let html = fs.readFileSync(resourcePath, 'utf-8');
    // vscode不支持直接加载本地资源，需要替换成其专有路径格式，这里只是简单的将样式和JS的路径替换
    html = html.replace(/(<link.+?href="|<script.+?src="|<img.+?src=")(.+?)"/g, (m, $1, $2) => {
        return $1 + vscode.Uri.file(path.resolve(dirPath, $2)).with({ scheme: 'vscode-resource' }).toString() + '"';
    });
    return html;
}

/**
 * 执行回调函数
 * @param {*} panel 
 * @param {*} message 
 * @param {*} resp 
 */
function invokeCallback(panel, message, resp) {
    console.log('回调消息：', resp);
    // 错误码在400-600之间的，默认弹出错误提示
    if (typeof resp == 'object' && resp.code && resp.code >= 400 && resp.code < 600) {
        util.showError(resp.message || '发生未知错误！');
    }
    panel.webview.postMessage({ cmd: 'vscodeCallback', cbid: message.cbid, data: resp });
}

/**
 * 存放所有消息回调函数，根据 message.cmd 来决定调用哪个方法
 */
const messageHandler = {
    getConfig(global, message) {
        const result = vscode.workspace.getConfiguration().get(message.key);
        invokeCallback(global.panel, message, result);
    },
    setConfig(global, message) {
        // 写入配置文件，注意，默认写入工作区配置，而不是用户配置，最后一个true表示写入全局用户配置
        vscode.workspace.getConfiguration().update(message.key, message.value, true);
        util.showInfo('修改配置成功！');
    }
};

module.exports = function (context) {

    context.subscriptions.push(vscode.commands.registerCommand('extension.demo.showWelcome', function () {
        const panel = vscode.window.createWebviewPanel(
            'testWelcome', // viewType
            "魔筷前端公告板", // 视图标题
            vscode.ViewColumn.One, // 显示在编辑器的哪个部位
            {
                enableScripts: true, // 启用JS，默认禁用
            }
        );
        let global = { panel };
        panel.webview.html = getWebViewContent(context, 'src/view/custom-welcome.html');
        panel.webview.onDidReceiveMessage(message => {
            if (messageHandler[message.cmd]) {
                messageHandler[message.cmd](global, message);
            } else {
                util.showError(`未找到名为 ${message.cmd} 回调方法!`);
            }
        }, undefined, context.subscriptions);
    }));

    const key = 'vscodePluginDemo.showTip';
    // 如果设置里面开启了欢迎页显示，启动欢迎页
    if (vscode.workspace.getConfiguration().get(key)) {
        console.log('如果设置里面开启了欢迎页显示，启动欢迎页');
        vscode.commands.executeCommand('extension.demo.showWelcome');
    }
    console.log('如果设置里面开启了欢迎页显示，启动欢迎页1');
};


src/view/custom-welcome.js
// const vscode = acquireVsCodeApi();
const callbacks = {};

/**
 * 调用vscode原生api
 * @param data 可以是类似 {cmd: 'xxx', param1: 'xxx'}，也可以直接是 cmd 字符串
 * @param cb 可选的回调函数
 */
function callVscode(data, cb) {
    if (typeof data === 'string') {
        data = { cmd: data };
    }
    if (cb) {
        // 时间戳加上5位随机数
        const cbid = Date.now() + '' + Math.round(Math.random() * 100000);
        callbacks[cbid] = cb;
        data.cbid = cbid;
    }
    // vscode.postMessage(data);
}

window.addEventListener('message', event => {
    const message = event.data;
    switch (message.cmd) {
        case 'vscodeCallback':
            console.log(message.data);
            (callbacks[message.cbid] || function () { })(message.data);
            delete callbacks[message.cbid];
            break;
        default: break;
    }
});

new Vue({
    el: '#app',
    data: {
        userName: '',
        time: '',
        show: true,
        dutyArr: ["余杰", "陈聪", "张晓旭", "徐文华", "薛纪宏", "林紫薇", "王瑞平", "佟刚", "夏创易", "雷涛", "张圆圆", "赵博文", "姚俊", "林海岳"],
        choosedList: []
    },
    mounted() {
        this.time = this.getTime();
        callVscode({ cmd: 'getConfig', key: 'vscodePluginDemo.yourName' }, userName => this.userName = userName);
        callVscode({ cmd: 'getConfig', key: 'vscodePluginDemo.showTip' }, show => this.show = show);
        this.todyIsWho()
    },
    watch: {
        show(nv, ov) {
            callVscode({ cmd: 'setConfig', key: 'vscodePluginDemo.showTip', value: nv }, null);
        }
    },
    methods: {
        todyIsWho() {
            let startDay = this.getTargetTime("2020-08-18 00:00:00");
            let dayTotalSeconds = 24 * 60 * 60 * 1000;
            let now = new Date();
            let lastTherrDay = now.getTime() - dayTotalSeconds * 4;
            let passDay = this.diffTime(now, '2020-8-18 00:00:00');
            let totalDays = passDay.days;
            let dutyArrLen = this.dutyArr.length;
            let choosedIndex = totalDays % dutyArrLen;
            let featureTenDay = [];
            console.log(choosedIndex);
            // console.log(this.dutyArr[choosedIndex]);

            // console.log(111, this.timeStampToDate(startDay));
            for (let i = 0; i < dutyArrLen; i++) {
                let time = now.getTime() + (dayTotalSeconds * i);
                let goodDay = this.timeStampToDate(time);
                let choosedMan = this.dutyArr[(choosedIndex + i) % dutyArrLen];
                goodDay['choosedMan'] = choosedMan;
                featureTenDay.push(goodDay)
                // console.log(featureTenDay);
            }
            this.choosedList = featureTenDay;
        },
        // t = "2014-08-08 08:08:08" 日期转为时间戳
        getTargetTime(t) {
            let d = t.split(" ")[0],
                h = t.split(" ")[1],
                date = new Date()

            date.setYear(d.split("-")[0])
            date.setMonth(d.split("-")[1] - 1)
            date.setDate(d.split("-")[2])

            date.setHours(h.split(":")[0])
            date.setMinutes(h.split(":")[1])
            date.setSeconds(h.split(":")[2])

            return date.getTime()
        },
        // 时间戳转为日期
        timeStampToDate(timeStamp) {
            let time = new Date(timeStamp);
            let year = time.getFullYear();
            let month = time.getMonth() + 1;
            let day = time.getDate();
            month = month < 10 ? ('0' + month) : month;
            day = day < 10 ? ('0' + day) : day;
            let week = "日一二三四五六".charAt(time.getDay());
            let obj = { year, month, day, week };
            // console.log(obj);
            return obj;
        },
        // console.log(diffTime(new Date(), '2019-8-19 16:00:00','Minutes'))
        diffTime(startDate, endDate, location) {
            let diff = new Date(endDate).getTime() - startDate;//时间差的毫秒数
            diff = Math.abs(diff)
            //计算出相差天数
            let days = Math.floor(diff / (24 * 3600 * 1000));

            //计算出小时数
            let leave1 = diff % (24 * 3600 * 1000);    //计算天数后剩余的毫秒数
            let hours = Math.floor(leave1 / (3600 * 1000));

            //计算相差分钟数
            let leave2 = leave1 % (3600 * 1000);        //计算小时数后剩余的毫秒数
            let minutes = Math.floor(leave2 / (60 * 1000));

            //计算相差秒数
            let leave3 = leave2 % (60 * 1000);      //计算分钟数后剩余的毫秒数
            let seconds = Math.round(leave3 / 1000);
            let returnStr = "";
            if (location === "Day") {
                return returnStr = "还有" + days + "天";
            } else if (location === "Hours") {
                return returnStr = "还有" + (hours + days * 24) + "小时";
            } else if (location === "Minutes") {
                return returnStr = "还有" + (minutes + (hours + days * 24) * 60) + "分钟";
            } else if (location === "Seconds") {
                return returnStr = "还有" + (seconds + (minutes + (hours + days * 24) * 60) * 60) + "秒";
            } else {
                // return returnStr = "还有" + days + "天" + hours + "小时" + minutes + "分钟" + seconds + "秒";
                return { days, hours, minutes, seconds };
            }
        },
        toggleShowTip() {
            this.show = !this.show;

        },
        getTime() {
            const hour = new Date().getHours();
            if (hour <= 8) return '早上';
            else if (hour < 12) return '上午';
            else if (hour < 14) return '中午';
            else if (hour < 18) return '下午';
            return '晚上';
        }
    }
});